### PR TITLE
Data module should serialize to maps and proper tagged types

### PR DESCRIPTION
**Pull request by hansl**
Depends on https://github.com/many-protocol/specification/pull/81

### CODE REVIEW

File: src/many-modules/src/_5_data/types.rs [hunk 1]
```diff
@@ -1,83 +1,566 @@
...
-pub enum DataValue {
-    #[n(0)]
-    Counter(#[n(0)] DataValueTypeCounter),
-    #[n(1)]
-    Gauge(#[n(0)] DataValueTypeGauge),
+#[derive(Clone, Debug, Default)]
+pub struct DataSet {
+    types: BTreeMap<DataIndex, DataInfo>,
+    values: BTreeMap<DataIndex, Arc<RwLock<DataValue>>>,
+}
```
Comment (fmorency) [inline src/many-modules/src/_5_data/types.rs:113]: Let's discuss this

File: src/many-modules/src/_5_data/types.rs [hunk 2]
```diff
@@ -1,83 +1,566 @@
...
-            DataValueTypeGauge::Float(_) => {
-                Err("Floats can't be converted to BigInt without loss".into())
+    pub fn as_gauge_int(&self) -> Option<&BigInt> {
+        match self {
+            Self::GaugeInt(v) => Some(v),
+            _ => None,
+        }
+    }
+
+    pub fn convert(&self, ty: DataType) -> Result<Self, ManyError> {
```
Comment (fmorency) [inline src/many-modules/src/_5_data/types.rs:351]: `convert_unknown`? or `try_into()`?

File: src/many-modules/src/_5_data/types.rs [hunk 3]
```diff
@@ -1,83 +1,566 @@
...
+            DataType::Counter => Self::Counter(d.u64()?),
+            DataType::GaugeInt => Self::GaugeInt(match d.datatype()? {
+                Type::U8
+                | Type::U16
+                | Type::U32
+                | Type::U64
+                | Type::I8
+                | Type::I16
+                | Type::I32
+                | Type::I64 => d.i64()?.into(),
```
Comment (fmorency) [inline src/many-modules/src/_5_data/types.rs:430]: It's not safe to read an `u64` as an `i64`

u64::MAX : 18_446_744_073_709_551_615u64
i64::MAX : 9_223_372_036_854_775_807i64

File: src/many-modules/src/_5_data/types.rs [hunk 4]
```diff
@@ -1,83 +1,566 @@
...
+    }
 }
 
 #[derive(Clone, Debug, Encode, Decode, Eq, PartialEq)]
 pub struct DataInfo {
     #[n(0)]
     pub r#type: DataType,
     #[n(1)]
-    pub shortname: String,
+    pub shortname: Cow<'static, str>,
```
Comment (fmorency) [inline src/many-modules/src/_5_data/types.rs:447]: Out of curiosity, why did you change this?

File: src/many-modules/src/_5_data/types.rs [hunk 5]
```diff
@@ -1,83 +1,566 @@
...
+    fn fails_on_unknown_index() {
+        let query_return = cbor_diag::parse_diag(r#"{ [0, [2, 3]]: 66051 }"#)
+            .unwrap()
+            .to_bytes();
+
+        let mut ds = DataSet::default().with_known_types().unwrap();
+
+        assert!(ds
+            .decode_merge(&mut Decoder::new(&query_return), false)
+            .is_err());
```
Comment (fmorency) [inline src/many-modules/src/_5_data/types.rs:520]: Check the error type

File: src/many-modules/src/_5_data/types.rs [hunk 6]
```diff
@@ -1,83 +1,566 @@
...
+        );
+
+        // This should be non-equal because it contains types that aren't in the
+        // original buffer.
+        let bytes = minicbor::to_vec(&ds).unwrap();
+        assert_ne!(&query_return, &bytes);
+
+        // Let's try again with the set of indices we have in the input buffer.
+        let bytes = minicbor::to_vec_with(&ds, &mut BTreeSet::from([TOTAL_COUNT_INDEX])).unwrap();
+        assert_eq!(&query_return, &bytes);
```
Comment (fmorency) [inline src/many-modules/src/_5_data/types.rs:564]: Add a test trying `to_vec_with` with an unsupported context.

File: src/many-types/src/cbor.rs [hunk 7]
```diff
@@ -1,28 +1,47 @@
...
     Bytes(Vec<u8>),
     Array(Vec<CborAny>),
     Map(BTreeMap<CborAny, CborAny>),
+    Tagged(Tag, Box<CborAny>),
+}
+
+impl CborAny {
+    pub fn as_u64(&self) -> Option<u64> {
+        match self {
+            Self::Int(i) if i >= &0 => Some(*i as u64),
```
Comment (fmorency) [inline src/many-types/src/cbor.rs:22]: That was confusing for a second.

**Review by fmorency — CHANGES_REQUESTED** [review]
Thanks!

Some minor changes/comments.

I want to discuss the DataSet thing before reviewing.

For some reason, the CI wasn't unable to upload to Codecov.

File: src/many-modules/src/_5_data/types.rs [hunk 8]
```diff
@@ -1,83 +1,573 @@
...
+use minicbor::{decode, encode, Decode, Decoder, Encode, Encoder};
+use num_bigint::{BigInt, Sign};
+use num_traits::cast::ToPrimitive;
+use std::borrow::Cow;
+use std::collections::{BTreeMap, BTreeSet};
+use std::sync::{Arc, PoisonError, RwLock, RwLockReadGuard};
 
-pub type DataIndex = AttributeRelatedIndex;
+macro_rules! decl_type {
+    ( ($index_name: ident, $info_name: ident) => ( $short: literal, $index: tt, gauge_int ) ) => {
```
[the remark this answers is not in the document]
Comment (hansl) [reply src/many-modules/src/_5_data/types.rs:22]: Added the missing one, actually.

File: src/many-modules/src/_5_data/types.rs [hunk 9]
```diff
@@ -1,83 +1,573 @@
...
+        Self(AttributeRelatedIndex::new(attribute))
+    }
+
+    #[inline]
+    pub const fn with_index(self, index: u32) -> Self {
+        Self(self.0.with_index(index))
+    }
+
+    #[inline]
+    const fn new_array<const N: usize>(arr: [u32; N]) -> Self {
```
[the remark this answers is not in the document]
Comment (hansl) [reply src/many-modules/src/_5_data/types.rs:76]: Refactored this code. Cheers.

File: src/many-modules/src/_5_data/types.rs (hunk 2 above)
Comment (hansl) [reply src/many-modules/src/_5_data/types.rs:351]: Removed the function. Now it's not needed.

File: src/many-modules/src/_5_data/types.rs (hunk 6 above)
Comment (hansl) [reply src/many-modules/src/_5_data/types.rs:530]: Wouldn't compile. I added a test with an unknown index though.

**Comment by hansl** [comment]
This is ready to review and merge.

**Review by fmorency — CHANGES_REQUESTED** [review]
The account count resiliency test is failing with this PR.

https://app.circleci.com/pipelines/github/liftedinit/many-rs/1989/workflows/4b44c598-f586-48c7-bf06-ebfa1c4d11ce/jobs/4964/tests#failed-test-0

**Comment by hansl** [comment]
What is going on? I've pushed a new commit to my fork and it's not showing up here.

**Review by fmorency — APPROVED** [review]
